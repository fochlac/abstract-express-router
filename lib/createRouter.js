import express, { Router } from 'express'

import bodyparser from 'body-parser'
import compression from 'compression'
import cookieParser from 'cookie-parser'
import { createLogger } from './utils/log'
import { createValidate } from './middleware/validate'

const methods = ['get', 'post', 'put', 'delete']
const reservedKeywords = [
	...methods,
	'params',
	'body',
	'middleware',
	'query',
	'controller',
	'static',
]

export function createRouter(
	routerConfig,
	{ logLevel, logger, useCatchAll = true, bodyParserOptions = {} } = {},
) {
	const log = createLogger({ logLevel, logger })
	const validate = createValidate(log)

	const router = express()

	router.use(cookieParser())
	router.use(bodyparser.json(bodyParserOptions))
	router.use(bodyparser.urlencoded({ extended: true, ...bodyParserOptions }))
	router.use(compression())

	router.use((req, res, next) => {
		log(4, `${req.method} call to ${req.originalUrl}`)
		next()
	})

	createSubrouter({ ...routerConfig }, log, validate, router)

	if (useCatchAll) {
		router.use((req, res) => {
			log(3, `Call to unknown path: ${req.path}`)
			res.status(404).send({ success: false, reason: 'Not_Found' })
		})
	}

	return router
}

function createSubrouter(config, log, validate, router = Router({ mergeParams: true })) {
	const { params, body, query, middleware } = config

	if (params || body || query) {
		router.use(validate({ params, body, query }))
	}

	if (middleware) {
		if (!middleware.forEach || middleware.some(mw => typeof mw !== 'function')) {
			log(1, 'Invalid router config. Middleware needs to be an array of functions.')
			throw new Error('Invalid router config. Middleware needs to be an array of functions.')
		}

		middleware.forEach(mw => router.use(mw))
	}

	methods.forEach(method => {
		if (config[method]) {
			createEndpoint(router, method, config[method], log, validate)
		}
	})

	Object.keys(config)
		.filter(key => !reservedKeywords.includes(key))
		.forEach(key => {
			router.use('/' + key, createSubrouter(config[key], log, validate))
		})

	if (config.static) {
		router.use(express.static(config.static))
	}

	return router
}

function createEndpoint(router, method, config, log, validate) {
	const { params, body, query, controller, middleware = [] } = config
	if (!controller || typeof controller !== 'function') {
		log(1, 'Invalid router config. Endpoints need to define a controller function.')
		throw new Error('Invalid router config. Endpoints need to define a controller function.')
	}
	if (!middleware.forEach || middleware.some(mw => typeof mw !== 'function')) {
		log(1, 'Invalid router config. Middleware needs to be an array of functions.')
		throw new Error('Invalid router config. Middleware needs to be an array of functions.')
	}
	router[method]('/', validate({ params, body, query }), ...middleware, controller)
}
