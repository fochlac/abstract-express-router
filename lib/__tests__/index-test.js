import { resetSpies, spies, testConfig } from './data/test-config'

import { createRouter } from '../createRouter'
import { spy } from 'sinon'
import supertest from 'supertest'

describe('integration test', () => {
	const logger = spy()
	const app = createRouter(testConfig, { logLevel: 4, logger })

	it('should throw for invalid routing configs', () => {
		try {
			createRouter({ middleware: ['test'] }, { logLevel: 1, logger })
		} catch (err) {
			expect(err.toString()).toEqual(
				'Error: Invalid router config. Middleware needs to be an array of functions.',
			)
		}
		try {
			createRouter(
				{ post: { middleware: ['test'], controller: Function.prototype } },
				{ logLevel: 1, logger },
			)
		} catch (err) {
			expect(err.toString()).toEqual(
				'Error: Invalid router config. Middleware needs to be an array of functions.',
			)
		}
		try {
			createRouter({ post: {} }, { logLevel: 1, logger })
		} catch (err) {
			expect(err.toString()).toEqual(
				'Error: Invalid router config. Endpoints need to define a controller function.',
			)
		}
	})

	it('should call root level middleware for any request', () => {
		resetSpies()
		logger.resetHistory()
		return supertest(app)
			.get('/api')
			.then(result => {
				expect(result.status).toEqual(404)
				expect(spies.testMiddleware.callCount).toEqual(1)
				expect(logger.args).toMatchSnapshot()
			})
	})

	it('should call correct middlewares, validation and controller for valid request to "/api/branch"', () => {
		resetSpies()
		logger.resetHistory()
		return supertest(app)
			.post('/api/branches')
			.send({ report: 'report', repository: 'foo', branch: '123' })
			.then(result => {
				expect(spies.testMiddleware.callCount).toEqual(1)
				expect(spies.testMiddlewareBranch.callCount).toEqual(1)
				expect(spies.addBranchReport.callCount).toEqual(1)
				expect(result.status).toEqual(200)
				expect(logger.args).toMatchSnapshot()
			})
	})

	it('should call correct middlewares and validation request to "/api/branch" with invalid body', () => {
		resetSpies()
		logger.resetHistory()
		return supertest(app)
			.post('/api/branches')
			.send({ report: 'report', repository: 'foo', branch: 'asd' })
			.then(result => {
				expect(spies.testMiddleware.callCount).toEqual(1)
				expect(spies.testMiddlewareBranch.callCount).toEqual(1)
				expect(spies.addBranchReport.callCount).toEqual(0)
				expect(result.status).toEqual(400)
				expect(logger.args).toMatchSnapshot()
			})
	})

	it('should skip to next route if path param is invalid', () => {
		resetSpies()
		logger.resetHistory()
		return supertest(app)
			.post('/api/pullrequests/asad')
			.send({ report: 'report', repository: 'foo', branch: 'asd' })
			.then(result => {
				expect(spies.testMiddleware.callCount).toEqual(1)
				expect(spies.testMiddlewareBranch.callCount).toEqual(0)
				expect(spies.addBranchReport.callCount).toEqual(0)
				expect(result.status).toEqual(404)
				expect(logger.args).toMatchSnapshot()
			})
	})

	it('should call full route if every parameter is valid', () => {
		resetSpies()
		logger.resetHistory()
		return supertest(app)
			.post('/api/pullrequests/123/')
			.send({ report: 'report', repository: 'foo', base: '123', pullRequestId: 123 })
			.then(result => {
				expect(spies.testMiddleware.callCount).toEqual(1)
				expect(spies.testMiddlewareBranch.callCount).toEqual(0)
				expect(spies.testMiddlewareEndPoint.callCount).toEqual(1)
				expect(spies.addBranchReport.callCount).toEqual(0)
				expect(spies.addPullRequestReport.callCount).toEqual(1)
				expect(result.status).toEqual(200)
				expect(logger.args).toMatchSnapshot()
			})
	})
})
