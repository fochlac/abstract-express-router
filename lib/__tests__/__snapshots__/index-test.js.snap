// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test should call correct middlewares and validation request to "/api/branch" with invalid body 1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "POST call to /api/branches",
      "requestId": 111,
    },
  ],
  Array [
    Object {
      "level": 2,
      "message": "Validation failed at field branch for call to /api/branches",
      "requestId": 111,
    },
  ],
]
`;

exports[`integration test should call correct middlewares, validation and controller for valid request to "/api/branch" 1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "POST call to /api/branches",
      "requestId": 74,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Validation succeded for call to /api/branches",
      "requestId": 74,
    },
  ],
]
`;

exports[`integration test should call full route if every parameter is valid 1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "POST call to /api/pullrequests/123/",
      "requestId": 222,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Validation succeded for call to /api/pullrequests/123",
      "requestId": 222,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Validation succeded for call to /api/pullrequests/123",
      "requestId": 222,
    },
  ],
]
`;

exports[`integration test should call root level middleware for any request 1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "GET call to /api",
      "requestId": 37,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Call to unknown path: /api",
      "requestId": 37,
    },
  ],
]
`;

exports[`integration test should call static middleware if it doesnt match anything else for gets 1`] = `Array []`;

exports[`integration test should return default 404 without catchAll and matching route  1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "POST call to /api/pullrequests/asad",
      "requestId": 185,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Validation of path parameters failed for call to /api/pullrequests/asad. Trying next route.",
      "requestId": 185,
    },
  ],
]
`;

exports[`integration test should skip to next route if path param is invalid 1`] = `
Array [
  Array [
    Object {
      "level": 4,
      "message": "POST call to /api/pullrequests/asad",
      "requestId": 148,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Validation of path parameters failed for call to /api/pullrequests/asad. Trying next route.",
      "requestId": 148,
    },
  ],
  Array [
    Object {
      "level": 3,
      "message": "Call to unknown path: /api/pullrequests/asad",
      "requestId": 148,
    },
  ],
]
`;
